linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1221,23 +1230,79 @@ private[optimizer] abstract class OptimizerCore(
         cont(PreTransLit(IntLiteral(resultValue)))
 
       case _ =>
-        resolveLocalDef(preTransQual) match {
-          case PreTransRecordTree(newQual, origType, cancelFun) =>
-            val recordType = newQual.tpe.asInstanceOf[RecordType]
-            val recordField = recordType.findField(field.name)
-            val sel = RecordSelect(newQual, field)(recordField.tpe)
-            sel.tpe match {
-              case _: RecordType =>
-                cont(PreTransRecordTree(sel, RefinedType(expectedType), cancelFun))
-              case _ =>
-                cont(PreTransTree(sel, RefinedType(sel.tpe)))
+        def default: TailRec[Tree] = {
+          resolveLocalDef(preTransQual) match {
+            case PreTransRecordTree(newQual, origType, cancelFun) =>
+              val recordType = newQual.tpe.asInstanceOf[RecordType]
+              val recordField = recordType.findField(field.name)
+              val sel = RecordSelect(newQual, field)(recordField.tpe)
+              sel.tpe match {
+                case _: RecordType =>
+                  cont(PreTransRecordTree(sel, RefinedType(expectedType), cancelFun))
+                case _ =>
+                  cont(PreTransTree(sel, RefinedType(sel.tpe)))
+              }
+
+            case PreTransTree(newQual, newQualType) =>
+              val newQual1 = maybeAssumeNotNull(newQual, newQualType)
+              cont(PreTransTree(Select(newQual1, className, field)(expectedType),
+                  RefinedType(expectedType)))
+          }
+        }
+
+        preTransQual.tpe match {
+          // Try to inline an inlineable field body
+          case RefinedType(ClassType(qualClassName), _, _) if !isLhsOfAssign =>
+            if (myself.exists(m => m.enclosingClassName == qualClassName && m.methodName.isConstructor)) {
+              /* Within the constructor of a class, we cannot trust the
+               * inlineable field bodies of that class, since they only reflect
+               * the values of fields when the instance is fully initialized.
+               */
+              default
+            } else {
+              inlineableFieldBodies(qualClassName).fieldBodies.get((className, field.name)) match {
+                case None =>
+                  default
+                case Some(fieldBody) =>
+                  val qualSideEffects = checkNotNullStatement(preTransQual)
+                  val fieldBodyTree = fieldBodyToTree(fieldBody)
```

Comment:
(non-actionable): It feels a bit odd that we construct an intermediate tree here only to pattern match again on it for pretransforming. But maybe that's a better option to keep the logic in one spot.